src/aspire/source/__init__.py

```diff
@@ -623,31 +626,38 @@ def save_metadata(
                             ]
                         )
 
-                # initial the star file object and save it
-                starfile = StarFile(blocks=[StarFileBlock(loops=[df])])
-                starfile.save(f)
+                mrcs_filenames = [
+                    df["_rlnImageName"][i].split("@")[1] for i in range(self.n)
+                ]
+
+            # initial the star file object and save it
+            starfile = StarFile(blocks=[StarFileBlock(loops=[df])])
+            starfile.save(f)
 
-        self._metadata_out = df
+        return mrcs_filenames
 
-    def save_images(self, starfile_filepath, batch_size=512, overwrite=False):
+    def save_images(
+        self, starfile_filepath, mrcs_filenames=None, batch_size=512, overwrite=False
+    ):
 
         """
         Save an ImageSource to MRCS files
 
         Note that .mrcs files are saved at the same location as the STAR file.
 
+        :param mrcs_filenames: Filename list for save all images
         :param starfile_filepath: Path to STAR file where we want to save image_source
         :param batch_size: Batch size of images to query from the `ImageSource` object.
             if `save_mode` is not `single`, images in the same batch will save to one MRCS file.
-        :param save_mode: Whether to save all images in a single or multiple files in batch size.
         :param overwrite: Whether to overwrite any .mrcs files found at the target location.
         :return: None
         """
 
+        if mrcs_filenames is None:
```

Comment:
If no filenames are specified, it should get them from the metadata. If there are no filenames there, it should error.

Reply:
Added to generate filename indices from metadata in commit 7d141bd.